### ISSUE TITLE
Empty immutable collections should be explicitly documented as singletons

### ISSUE THREAD

**Issue by eamonnmcmanus** [google/guava#2083]
`ImmutableList.of()` says this:

> Returns the empty immutable list. This set behaves and performs comparably to `Collections.emptyList()`, and is preferable mainly for consistency and maintainability of your code.

The use of the phrase "the empty immutable list" suggests that this method always returns the same object, as indeed it does. But this is implied only by the use of the definite article and the reference to `Collecitons.emptyList()`. It would be better to state it more explicitly. By comparison, `Collections.emptyList()` itself says this:

> Implementation note: Implementations of this method need not create a separate `List` object for each call. Using this method is likely to have comparable cost to using the like-named field. (Unlike this method, the field does not provide type safety.)

Of course the same applies to the other empty-collection methods.

**Comment by kevinb9n** [comment]
Yes, this seems worth noting. We use `<b>Performance note:</b> ...` for things like this.

**Comment by cjosw** [comment]
I have raised a pull request for this: https://github.com/google/guava/pull/2247

**Comment by bedla** [comment]
Question: Why is ImmutableList.of() preferable to Collections.emptyList() ? JavaDoc: "for consistency and maintainability of your code" does not clarify it, imo.

**Comment by Stephan202** [comment]
The return type of `ImmutableList.of()` is an `ImmutableList`. Guava recommends using such `Immutable*` types throughout, because this way immutability is communicated and propagated at the level of the type system. Not only does this increase maintainability (because it makes it easier to reason about the code), the compiler will then also warn about code that invokes a mutator method on such types. (Because they're marked deprecated.)

**Comment by bedla** [comment]
I understand what you mean, but it works only when you have ImmutableList at signatures, but when you prefer coding against interface, you have List there...

**Comment by ronshapiro** [comment]
We treat ImmutableList as an interface in that sense. It has meaningful
semantics that are frequently relevant for a caller to know. In the same
way that you can return a Collection vs. a Set, even though Set doesn't
have any different methods, it is clearly a different type.

**Comment by bedla** [comment]
Ok, i see :] thanks a lot

**Comment by cpovirk** [comment]
I'm now imagining that we'll do this, only to find that inline classes will eventually force us to make it _not_ a singleton (though I imagine we'd still try to have only one instance for each possible specialization) :) If we wanted, we could hedge our bets with something like [what the Javadoc for `emptyList` says](https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/util/Collections.html#emptyList()). Of course, that's not so much "hedging our bets" as it is "refusing to guarantee the thing that we want people to count on" :) And as Éamonn pointed out, our current phrasing already technically implies that it's a singleton. And really, if inline classes arrive with widespread enough implications to make our empty immutable collections _not_ be singletons, I would expect for users to have bigger concerns than what our doc said :) And probably they'll be understanding (so long as we're not somehow forced to create a new instance for _every_ call, which seems unlikely).

**Comment by cpovirk** [comment]
(Oh, and I was also going to note that we could conceivably make similar changes for `Optional.absent` and quite possibly other types.)
